runtime/new-recipe.js

```diff
@@ -15,30 +15,106 @@ class Connection extends Edge {
 }
 
 class Particle extends Node {
-  get id() {} // Not resolved until we have an ID.
-  get name() {}
-  get tags() {}
-  get providedSlots() {} // Slot*
-  get consumedSlots() {} // SlotConnection*
-  get parameters() {} // {parameter -> ViewConnection}
+  constructor(recipe, name) {
+    assert(recipe);
+    this._recipe = recipe;
+    this._id = undefined;
+    this._name = name;
+    this._tags = [];
+    this._providedSlots = [];
+    this._consumedSlots = [];
+    this._connections = {};
+    this._unnamedConnections = [];
+  }
+  get id() { return this._id; } // Not resolved until we have an ID.
+  get name() { return this._name; }
+  get tags() { return this._tags; }
+  get providedSlots() { return this._providedSlots; } // Slot*
+  get consumedSlots() { return this._consumedSlots; } // SlotConnection*
+  get connections() { return this._connections; } // {parameter -> ViewConnection}
+
+  addUnnamedConnection() {
+    var connection = new ViewConnection(undefined, this);
+    this._unnamedConnections.push(connection);
+    return connection;
+  }
+
+  addConnectionName(name) {
+    this._connections[name] = new ViewConnection(name, this);
+    return this._connections[name];
+  }
+
+  nameConnection(connection, name) {
+    var idx = this._unnamedConnections.indexOf(connection);
+    assert(idx >= 0);
+    connection._name = name;
+    this._connections[name] = connection;
+    this._unnamedConnections.splice(idx, 1);
+  }
+
+  isResolved() {
+    return this._unnamedConnections.length == 0;
+    // and probably some other stuff too.
+  }
 }
 
 class View extends Node {
+  constructor(recipe) {
+    assert(recipe);
+    this._recipe = recipe;
+    this._id = undefined;
+    this._tags = [];
+    this._type = undefined;
+    this._create = false;
+    this._connections = [];
+  }
   // a resolved View has either an id or create=true
-  get tags() {} // only tags owned by the view
-  get type() {} // nullable
-  get id() {}
-  get create() {}
-  get connections() {} // ViewConnection*
+  get tags() { return this._tags; } // only tags owned by the view
+  get type() { return this._type; } // nullable
+  get id() { return this._id; }
+  get create() { return this._create; }
+  get connections() { return this._connections } // ViewConnection*
+
+  isResolved() {
+    return this._id !== undefined || this._create == true;
+  }
 }
 
 class ViewConnection extends Connection {
-  get name() {} // Parameter name?
-  get tags() {}
-  get type() {}
-  get direction() {} // in/out
-  get view() {} // View?
-  get particle() {} // never null
+  constructor(name, particle) {
```

Comment:
These classes aren't exported directly so they aren't easily constructable. I probably should write some documentation on usage of this module, but I might wait until it's bedded down a bit more.